### PR TITLE
Iou two lists

### PR DESCRIPTION
**Pull request by Renkai**
This pr solves https://github.com/eto-ai/rikai/issues/339
I use the name `iou_matrix` not `iou` because the python interpreter believes with name `iou` the type would be ambiguous with another function https://github.com/eto-ai/rikai/blob/38457fc7657ab46d815a5f1488f8c55aab5e7012/python/rikai/types/geometry.py#L302

### CODE REVIEW

**Comment by Renkai** [comment]
@eddyxu code updated, not sure if it is the right vectorized version because I'm new to vectorizing, the benchmark is being prepared but you can have a look at the code in advance if you'd like.

**Comment by Renkai** [comment]
```
$ python python/benchmark/types/iou_benchmark.py 1000 1000 10

naive method cost 15.941821 seconds
vectorized method cost 0.360174 seconds
naive/vectorized 44.26144308028897
```
From my benchmark, the vectorized version is 44 times faster than the naive version for two parameters with length 1000.

File: python/benchmark/types/iou_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import datetime
```
Comment (eddyxu) [inline python/benchmark/types/iou_benchmark.py:1]: Good work! Could you add Apache license ?

Also run `isort` to put the imports into different sections?

File: python/rikai/types/geometry.py [hunk 2]
```diff
@@ -281,6 +281,36 @@ def _area(arr: np.ndarray) -> np.array:
             0, arr[:, 3] - arr[:, 1]
         )
 
+    @staticmethod
+    def iou_matrix(box_list1: Union[Sequence[Box2d], np.ndarray],
```
Comment (eddyxu) [inline python/rikai/types/geometry.py:285]: lets rename them as `boxes1, boxes2` ? they are not necessary a `list` ;)

File: python/benchmark/types/iou_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        x_max = random.uniform(x_min, 1)
+        y_max = random.uniform(y_min, 1)
+        return Box2d(x_min, y_min, x_max, y_max)
+
+    list1 = [a_random_box2d() for _ in range(0, list1_len)]
+    list2 = [a_random_box2d() for _ in range(0, list2_len)]
+
+    time0 = datetime.now()
+    count = 0
+    while count < times:
```
Comment (eddyxu) [inline python/benchmark/types/iou_benchmark.py:42]: can we use [timeit](https://docs.python.org/3/library/timeit.html) here?

File: python/rikai/types/geometry.py [hunk 4]
```diff
@@ -281,6 +281,36 @@ def _area(arr: np.ndarray) -> np.array:
...
+
+        xmin = np.maximum(box_list1[:, 0].reshape((row_count, -1)), box_list2[:, 0])
+        ymin = np.maximum(box_list1[:, 1].reshape((row_count, -1)), box_list2[:, 1])
+        xmax = np.minimum(box_list1[:, 2].reshape((row_count, -1)), box_list2[:, 2])
+        ymax = np.minimum(box_list1[:, 3].reshape((row_count, -1)), box_list2[:, 3])
+
+        inter_area = np.maximum(0, xmax - xmin) * np.maximum(
+            0, ymax - ymin)
+
+        iou_mat = inter_area / (area1 + area2 - inter_area)
```
Comment (eddyxu) [inline python/rikai/types/geometry.py:310]: do we need to handle `DivideByZero` here?

File: python/rikai/types/geometry.py [hunk 5]
```diff
@@ -281,6 +281,36 @@ def _area(arr: np.ndarray) -> np.array:
             0, arr[:, 3] - arr[:, 1]
         )
 
+    @staticmethod
+    def iou_matrix(box_list1: Union[Sequence[Box2d], np.ndarray],
+                   box_list2: Union[Sequence[Box2d], np.ndarray]) -> np.ndarray:
+        """Compute intersection over union(IOU).
```
Comment (eddyxu) [inline python/rikai/types/geometry.py:287]: We use [numpy doc style](https://numpydoc.readthedocs.io/en/latest/format.html) in rikai. could you help to update the document to be aligned with the docstring style?

File: python/benchmark/types/iou_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    time1 = datetime.now()
+    count = 0
+    while count < times:
+        count += 1
+        Box2d.iou_matrix(list1, list2)
+
+    time2 = datetime.now()
+    naive_seconds = (time1 - time0).total_seconds()
+    vectorized_seconds = (time2 - time1).total_seconds()
+    print("naive method cost {} seconds".format(naive_seconds))
```
Comment (eddyxu) [inline python/benchmark/types/iou_benchmark.py:55]: Nice, wonder what does the result look like

**Comment by eddyxu** [comment]
Hey, @Renkai this is a nice work!

Can you help to report the numbers where the length of boxes are 10, 100, 1000? Also let us know your CPU model?

It is not mandate, but if possible, can we also get the same benchmark against MKL and openBlas (including MKL and openBlas version)?

For example, you can run

```python
>>> import numpy as np
>>> np.__config__.show()
```

to see what library is used.

File: python/rikai/types/geometry.py (hunk 2 above)
Comment (eddyxu) [reply python/rikai/types/geometry.py:285]: Shall we just renames this as `ious`, and allow the function takes `Union[Sequence[Box2d], np.ndarray, Box2d]`?

File: python/rikai/types/geometry.py (hunk 4 above)
Comment (Renkai) [reply python/rikai/types/geometry.py:340]: The default behavior is to print a runtime warning and gives `Nan` in the corresponding cell as a result, I believe it's an ok solution.

**Comment by Renkai** [comment]
```
In [1]: import numpy as np

In [2]: np.__config__.show()
blas_mkl_info:
  NOT AVAILABLE
blis_info:
  NOT AVAILABLE
openblas_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
blas_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
lapack_mkl_info:
  NOT AVAILABLE
openblas_lapack_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
lapack_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
```
Seems there is no `MKL` on my computer.

CPU Info
```
➜  ~ sysctl -n machdep.cpu.brand_string
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
```

With the box list length of 10/100/1000, the result of time cost division is 1.79/19.42/42.57
```
(rikai-example) ➜  rikai git:(iou_two_lists) ✗ python python/benchmark/types/iou_benchmark.py 10 10 10
naive method cost 0.002192802000000049 seconds
vectorized method cost 0.0012242569999999509 seconds
naive/vectorized 1.7911288234415954
(rikai-example) ➜  rikai git:(iou_two_lists) ✗ python python/benchmark/types/iou_benchmark.py 100 100 10
naive method cost 0.163356304 seconds
vectorized method cost 0.008413267999999974 seconds
naive/vectorized 19.41651020744858
(rikai-example) ➜  rikai git:(iou_two_lists) ✗ python python/benchmark/types/iou_benchmark.py 1000 1000 10
naive method cost 16.681013051 seconds
vectorized method cost 0.391952670000002 seconds
naive/vectorized 42.55874325591382
```

**Review by eddyxu — APPROVED** [review]
Looks nice. Thanks @Renkai

**Pull request merged**